include/seqan3/search/dream_index/interleaved_bloom_filter.hpp

```diff
@@ -0,0 +1,501 @@
+    interleaved_bloom_filter(interleaved_bloom_filter &&) = default; //!< Defaulted.
+    interleaved_bloom_filter & operator=(interleaved_bloom_filter &&) = default; //!< Defaulted.
+    ~interleaved_bloom_filter() = default; //!< Defaulted.
+
+    /*!\brief Construct an uncompressed Interleaved Bloom Filter.
+     * \param bins_ The number of bins.
+     * \param bin_size__ The bitvector size.
+     * \param funs The number of hash functions. Default 2.
+     *
+     * \attention This constructor can only be used to construct **uncompressed** Interleaved Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/dream_index/interleaved_bloom_filter_constructor.cpp
+     */
+    interleaved_bloom_filter(seqan3::bin_count bins_,
+                             seqan3::bin_size bin_size__,
+                             seqan3::hash_function_count funs = seqan3::hash_function_count{2u})
+    //!\cond
+        requires data_layout_mode == data_layout::uncompressed
+    //!\endcond
+    {
+        bins = bins_.get();
+        bin_size_ = bin_size__.get();
+        hash_funs = funs.get();
+
+        if (bins == 0)
+            throw std::logic_error{"The number of bins must be > 0."};
+        if (hash_funs == 0)
+            throw std::logic_error{"The number of hash functions must be > 0."};
+        if (bin_size_ == 0)
+            throw std::logic_error{"The size of a bin must be > 0."};
+
+        bin_words = (bins + 63) >> 6; // = ceil(bins/64)
+        technical_bins  = bin_words << 6; // = bin_words * 64
+        data = sdsl::bit_vector(technical_bins * bin_size_);
+
+        result_buffer.resize(bins);
+        hash_seeds.resize(hash_funs);
+
+        for (size_t i = 0; i < hash_funs; ++i)
+            hash_seeds[i] = i ^ 0x90b45d39fb6da1fa;
+    }
+
+    /*!\brief Construct a compressed Interleaved Bloom Filter.
+     * \param[in] ibf The uncompressed seqan3::interleaved_bloom_filter.
+     *
+     * \attention This constructor can only be used to construct **compressed** Interleaved Bloom Filters.
+     *
+     * \details
+     *
+     * ### Example
+     *
+     * \include test/snippet/search/dream_index/interleaved_bloom_filter_constructor_compressed.cpp
+     */
+    interleaved_bloom_filter(interleaved_bloom_filter<data_layout::uncompressed> const & ibf)
+    //!\cond
+        requires data_layout_mode == data_layout::compressed
+    //!\endcond
+    {
+        std::tie(bins, technical_bins, bin_size_, bin_words, hash_funs, hash_seeds) =
+            std::tie(ibf.bins, ibf.technical_bins, ibf.bin_size_, ibf.bin_words, ibf.hash_funs, ibf.hash_seeds);
+
+        data = sdsl::sd_vector<>{ibf.data};
+        result_buffer.resize(bins);
+    }
+    //!\}
+
+    /*!\name Modifiers
+     * \{
+     */
+    /*!\brief Inserts a value into a specific bin.
+     * \param[in] h   The raw hash value to process.
```

Comment:
Since we do not require a hash value here but an arbitrary number, I think we decided to name it like that?
```suggestion
     * \param[in] value   The raw numeric value to process.
```
Same for all places except in hash_and_fit